hnn_core/extracellular.py

```diff
@@ -0,0 +1,601 @@
+        self._nrn_r_transfer = None
+        self._nrn_times = None
+        self._nrn_voltages = None
+
+    def _build(self, cvode=None):
+        """Assemble NEURON objects for calculating extracellular potentials.
+
+        The handler is set up to maintain a vector of membrane currents at at
+        every inner segment of every section of every cell on each CVODE
+        integration step. In addition, it records a time vector of sample
+        times.
+
+        Parameters
+        ----------
+        cvode : instance of h.CVode
+            Multi order variable time step integration method.
+        """
+        # ordered list of h.Sections on this rank (if running in parallel)
+        secs_on_rank = _get_sections_on_this_rank()
+        # np.array of number of segments for each section, ordered as above
+        segment_counts = np.array(_get_segment_counts(secs_on_rank))
+        n_total_segments = segment_counts.sum()
+
+        # pointers assigned to _ref_i_membrane_ at each EACH internal segment
+        self._nrn_imem_ptrvec = h.PtrVector(n_total_segments)
+        # placeholder into which pointer values are read on each sim time step
+        self._nrn_imem_vec = h.Vector(n_total_segments)
+
+        ptr_count = 0
+        for sec in secs_on_rank:
+            for seg in sec:  # section end points (0, 1) not included
+                # set Nth pointer to the net membrane current at this segment
+                self._nrn_imem_ptrvec.pset(
+                    ptr_count, sec(seg.x)._ref_i_membrane_)
+                ptr_count += 1
+        if ptr_count != n_total_segments:
+            raise RuntimeError(f'Expected {n_total_segments} imem pointers, '
+                               f'got {ptr_count}.')
+
+        # transfer resistances for each segment (keep in Neuron Matrix object)
+        self._nrn_r_transfer = h.Matrix(self.n_contacts, n_total_segments)
+
+        for row, pos in enumerate(self.array.positions):
+            if self.array.method is not None:
+                transfer_resistance = list()
+                for sec in secs_on_rank:
+                    this_xfer_r = _transfer_resistance(
+                        sec, pos, sigma=self.array.sigma,
+                        method=self.array.method,
+                        min_distance=self.array.min_distance)
+                    transfer_resistance.extend(this_xfer_r)
+
+                self._nrn_r_transfer.setrow(row, h.Vector(transfer_resistance))
+            else:
+                # for testing, make a matrix of ones
+                self._nrn_r_transfer.setrow(row,
+                                            h.Vector(segment_counts.sum(), 1.))
+
+        # record time for each array
+        self._nrn_times = h.Vector().record(h._ref_t)
+
+        # contributions of all segments on this rank to total calculated
+        # potential at electrode (_PC.allreduce called in _simulate_dipole)
+        self._nrn_voltages = h.Vector()
```

Comment:
could we declare the shape here? It's really curious how it's used below. I thought these functioned like numpy arrays but there is an `append` method. Maybe just add as an inline comment if it's not possible using Neuron API

Reply:
Sadly: no. The `h.Vector.append` method is indeed a curious beast: it must be doing some sort of `realloc` on each iteration! Sounds terribly inefficient, and in C-terms it probably is. The reason the declaration isn't possible is that we don't know how many samples we'll get before the simulation is done.

Reply:
but in `CellResponse` we instantiate the `times` and also in `Dipole` ...